lib/mixlib/shellout.rb

```diff
@@ -199,9 +202,48 @@ def uid
 
     # The gid that the subprocess will switch to. If the group attribute is
     # given as a group name, it is converted to a gid by Etc.getgrnam
+    # If a group is not provided but with_logon is true, the user's primary
+    # group will be used.
     def gid
-      return nil unless group
-      group.kind_of?(Integer) ? group : Etc.getgrnam(group.to_s).gid
+      return group.kind_of?(Integer) ? group : Etc.getgrnam(group.to_s).gid if group 
+      return Etc.getpwuid(uid).gid if (with_logon && user)
```

Comment:
`validate_options` is being called only when variables are set using the options hash (`parse_options`).
I couldn't understand if `ShellOut` is intended to be used otherwise (e.g. instantiating blank class and populating variables later via attributes). 
If it is possible, `validate_options` isn't called every time, so it can't be relied upon to ensure the object being "stable", and we have to do that check inside the methods.
If it isn't, we should make these attributes read only (e.g. `attr_accessor :with_logon` > `attr_reader :with_logon`), and rely on `validate_options` being called.